### ISSUE TITLE
Mimirtool remote-read export fails with "out ouf bounds"

### ISSUE THREAD

**Issue by sysedwinistrator** [grafana/mimir#6466]
#### Describe the bug

I'm trying to export some metrics via `mimirtool remote-read export`. After it logs `Store TSDB blocks in './tsdb'`, it fails with this error message (some labels and values redacted):

```
INFO[0000] log level set to debug                       
INFO[0000] Created remote read client using endpoint 'http://localhost:8080/prometheus/api/v1/read' 
INFO[0000] Using existing TSDB in path './tsdb'         
INFO[0000] Querying time from=2023-10-23T12:23:54+02:00 to=2023-10-23T13:23:54+02:00 with selector={gcp_project_id="redacted",job="k8s-nodes"} 
INFO[0003] Store TSDB blocks in './tsdb'
mimirtool: error: process blocks: add sample for metric={__name__="apiserver_audit_requests_rejected_total",a_redacted_label="value",another_redacted_label="value"} ts=2023-10-23T12:00:05.815+02:00 value=0.000000: out of bounds, try --help
```

#### To Reproduce

Run the following command using mimirtool 2.10.3:

```
$ mimirtool --log.level=debug remote-read export --read-timeout=1h --id redacted --selector '{gcp_project_id="redacted",job="k8s-nodes"}' --address http://localhost:8080 --remote-read-path /prometheus/api/v1/read --tsdb-path ./tsdb
```

This happens with different selectors, tenants, and time windows. Although for some selectors, especially those returning fewer metrics, it does not happen.

#### Expected behavior

It looks like the metric sample for which the error message is printed is outside the time window of the query time logged earlier. If that's the case, the TSDB write code should probably just drop samples outside of the time frame. Or better, the remote read endpoint shouldn't return samples outside of the specified time window - although this is outside of this project's control for remote read endpoints other than Mimir.

It would also be useful if the debug log level would actually have an effect and print the file/line number of the code producing the message. 

#### Environment

- Infrastructure:
  - Mimir 2.10.3 running on Kubernetes (GKE) 
- Deployment tool:
  - mimir-distributed Helm chart version 5.1.2

**Comment by pracucci** [comment]
I see two possible causes:

1. The remote backend really returns a sample outside the requested time range
2. There's a bug in the remote read client returning a corrupted sample (not that we're aware of such issue, but it's an option)

In case (1) I agree we should just filter out (with a warning) the received samples outside of the time range, before trying to append them to TSDB. However, it could be useful if you could ensure that the sample we're receiving from the remote backend is outside the requested range, to ensure we're not in the case (2).

That being said, after confirming it, would you feel comfortable opening a PR for (1)?

**Comment by sysedwinistrator** [comment]
Sorry for the late reply. The function where the error message is coming from, `CreateBlocks` in `pkg/mimirtool/backfill/backfill.go`, should already skip samples that are outside the current block's range: https://github.com/grafana/mimir/blob/52c39fb6ce68442128f6784c5993a5ee09c8627f/pkg/mimirtool/backfill/backfill.go#L121-L123

I also played around with tests based on TeastEarlyCommit in `pkg/mimirtool/commands/remote_read_test.go` that create samples outside the time range, but they still passed. **EDIT**: probably because this has already been on main.

Since the `add sample for metric={__name__="apiserver_audit_requests_rejected_total",a_redacted_label="value",another_redacted_label="value"} ts=2023-10-23T12:00:05.815+02:00 value=0.000000: out of bounds` part of error message comes from Prometheus's TSDB library, I looked into that. While I found out that [ErrOutOfBounds](https://github.com/prometheus/prometheus/blob/b8f5b314aeddc140c5566eb145997655263bee5e/storage/interface.go#L35-L36) is "https://github.com/prometheus/prometheus/blob/b8f5b314aeddc140c5566eb145997655263bee5e/storage/interface.go#L35-L36", it doesn't really make sense here because the metrics returned from the remote read should be in order(?).

Anyway, while looking around in the promtool backfill code from which the mimirtool backfill code was originally copied, I found that after the code was added to mimirtool, a fix for a very similar issue has been added:
Issue: https://github.com/prometheus/prometheus/issues/8476
MR: https://github.com/prometheus/prometheus/pull/8496

The problem was that the block writer only allows appending samples that are at most half a block size older than the most recent sample appended so far (see code comment). The solution was to simply double the block duration when creating the block writer.

The same issue has already been adressed for mimirtool in this MR: https://github.com/grafana/mimir/pull/5700.
In this case the fix was to half the time increases in the outer loop and in the calculation of tsUpper, the maximum timestamp for sample.
I don't think the authors of the Mimir MR were aware of the Prometheus bug report and fix, as there were no references to that.
Anyway, this fix was not yet included in Mimir 2.10.3, but it is in Mimir 2.11 (which I just released after writing all this text).

**TL;DR:** Issue has been fixed in Prometheus in 2021, and more recently in Mimir (using a slightly different approach).
